### PR TITLE
tag filtering query

### PR DESCRIPTION
**Pull request by InformatiQ**
https://github.com/grafana/grafana/issues/4824
by using filtering the tags in the sql query we ensure that the issue of 'LIMIT 1000' does not affect the result (much)
it will still only show less 1000 (unique)dashboards in the case where the tag has 1000+ dashboards

### CODE REVIEW

File: pkg/services/sqlstore/dashboard.go
Comment (torkelo) [inline pkg/services/sqlstore/dashboard.go]: what is this doing here? And wont this require aggregation functions for fields in the select clause, and cause issue with the tag outer join so we do not get all the tags any more?

Comment (InformatiQ) [inline pkg/services/sqlstore/dashboard.go]: True, i missed that. I'll give it another shot

**Comment by InformatiQ** [comment]
could someone look help explain why th etests are failing?

**Comment by iksaif** [comment]
@InformatiQ  I believe it fails because you need to run gofmt on your files

File: pkg/services/search/handlers.go
Comment (iksaif) [inline pkg/services/search/handlers.go]: Just remove the code

File: pkg/services/sqlstore/dashboard.go
Comment (iksaif) [inline pkg/services/sqlstore/dashboard.go]: Remove this, or add a sentence to explain why it is here.

File: pkg/services/sqlstore/dashboard.go
Comment (iksaif) [inline pkg/services/sqlstore/dashboard.go]: ) GROUP BY

**Comment by torkelo** [comment]
could you write down the complete SQL query the code generates depending on different factors? There has be a clearner way to do this. I know we have to have sub queries to do the tag join & filter, but the way it's built here looks unnecessarily complex

**Review by iksaif — CHANGES_REQUESTED** [review]
Did you check that it works on PostgreSQL and sqlite too ?

**Comment by InformatiQ** [comment]
The change is only related to tags
if the tags are specified in the query list the query becomes this

SELECT dashboard.id, dashboard.title, dashboard.slug, dashboard_tag.term 
FROM 
    ( SELECT dashboard.id, dashboard.title, dashboard.slug, dashboard_tag.term 
      FROM dashboard 
      LEFT OUTER JOIN dashboard_tag 
          ON dashboard_tag.dashboard_id = dashboard.id 
      WHERE dashboard.org_id=1 
          AND dashboard_tag.term IN ( tags ) 
      GROUP BY dashboard.slug 
      HAVING COUNT(dashboard.slug) >= len(tags)
     ) as dashboard 
INNER JOIN dashboard_tag on dashboard.id = dashboard_tag.dashboard_id 
ORDER BY dashboard.title ASC 
LIMIT 1000;

Else it remains as it was before

File: pkg/services/sqlstore/dashboard.go
Comment (InformatiQ) [inline pkg/services/sqlstore/dashboard.go]: i am closing an earlier parenthese

Comment (iksaif) [inline pkg/services/sqlstore/dashboard.go]: I meant that you should add a space here

**Comment by daniellee** [comment]
Run `gofmt -s -d ./pkg` to see what gofmt is complaining about.

**Comment by InformatiQ** [comment]
tested with mysql and sqlite

**Comment by eddawley** [comment]
I believe this query will fail in postgresql since the subselect is pulling columns not in GROUP BY.  This is non-ANSI compliant behavior that is allowed in mysql/sqlite but I don't think in postgres.  Also the join should be an INNER JOIN.

You'll probably need to go with something like:

```
SELECT dashboard.id, dashboard.title, dashboard.slug, dashboard_tag.term
FROM
( SELECT dashboard.id
FROM dashboard
INNER JOIN dashboard_tag
ON dashboard_tag.dashboard_id = dashboard.id
WHERE dashboard.org_id=1
AND dashboard_tag.term IN ( tags )
GROUP BY dashboard.id
HAVING COUNT(dashboard.id) >= len(tags)
) as ids
INNER JOIN dashboard ON ids.id = dashboard.id
INNER JOIN dashboard_tag on dashboard.id = dashboard_tag.dashboard_id
ORDER BY dashboard.title ASC
LIMIT 1000;
```

**Comment by InformatiQ** [comment]
I could use some help with the failing test

**Comment by iksaif** [comment]
```
  * /home/ubuntu/.go_workspace/src/github.com/grafana/grafana/pkg/services/search/handlers_test.go 
  Line 54:
  Expected: '2'
  Actual:   '6'
  (Should be equal)
```

Could you try outputing the results to see what's wrong ?

**Comment by InformatiQ** [comment]
it actually makes sense to fail.
prior to my patch the tag filtering was hapenning in the code i removed from pkg/services/search/handlers.go 
so now the test is returning the full Hit:List as we are not doing the filtering here

File: pkg/services/sqlstore/dashboard.go
Comment (eddawley) [inline pkg/services/sqlstore/dashboard.go]: This query is now performing a logical OR on the tags whereas the original one was doing and AND (ie all tags were required).  Search in grafana appears to be ANDs so I think this is incorrect now.

Comment (InformatiQ) [inline pkg/services/sqlstore/dashboard.go]: the AND is done doing an OR and counting the occurences (see line 150) 
ROUP BY dashboard.id HAVING COUNT(dashboard.slug) >= ?)

Comment (eddawley) [inline pkg/services/sqlstore/dashboard.go]: Sorry, my mistake.  My quick look just missed that part.

**Comment by InformatiQ** [comment]
@torkelo is it better now?

**Comment by iksaif** [comment]
Looks good to me :).

**Review by iksaif — APPROVED** [review]

**Comment by iksaif** [comment]
@torkelo / @eddawley  what do you think ?

**Comment by torkelo** [comment]
Is it tested on MySQL porstgtand sqlite? If so this is mergable :)

**Comment by InformatiQ** [comment]
now it should work on the 3 DBs

**Comment by torkelo** [comment]
@daniellee think we should look at merging this tomorrow, as it will otherwise cause big melee conflicts with your dashboardnl folders work

**Comment by daniellee** [comment]
Will merge this after the 4.3.0 release this week so that it will be included in the 4.4.0 release. Looking forward to that merge conflict on the dashboard_folders branch :smiley:

**Comment by Thib17** [comment]
Hello @daniellee, since 4.3.0 is released, could we merge this PR ?

**Comment by daniellee** [comment]
@Thib17 It is in the milestone for 4.4.0 so I will be looking at merging this in the next week or two. I have to test it on all the databases first and see how much it conflicts with coming changes (the search ux is being redone and we are introducing a folders for dashboards).

**Comment by InformatiQ** [comment]
@daniellee any update on when this is getting merged? i see it is in the 5.0 milestone
could you give an estimate of the time?

**Comment by torkelo** [comment]
It's already in develop branch, but that branch needs a bit more work before we can merge, maybe mid to late August

**Comment by torkelo** [comment]
parts of this PR was taken and merged manually with the changes in the develop branch, the search SQL was changed alot with the addition of dashboard folders & permission checks. 

Anyway, closing this. The deveop branch will be merged to master hopefully with in 3-4 weeks.

**Comment by iksaif** [comment]
What's up with this ? Looks like the bug is still present in the latest version (and this makes search more or less unusable as soon as you have more than 1000 dashboards)

**Pull request closed without merging**
